lib/spack/spack/directory_layout.py

```diff
@@ -356,17 +355,22 @@ def check_activated(self, spec, ext_spec):
     def extension_file_path(self, spec):
         """Gets full path to an installed package's extension file"""
         _check_concrete(spec)
-        return os.path.join(self.layout.metadata_path(spec),
-                            self.extension_file_name)
+        if os.path.abspath(self.root) in os.path.abspath(spec.prefix):
```

Comment:
Use `prefix.startswith(root + os.path.sep)` instead of `in`, as `in` is just "is a substring of" and could fail in weird ways.  The root `/bar/baz` is "in" `/foo/bar/baz` but `/foo/bar/baz` is not a subdirectory.

Also the way `in` reads is confusing -- you want to know whether the prefix is a subdirectory of the root (i.e., it is "in" the root) but you are testing whether root is a substring of prefix (i.e., it is "in" the prefix -- which sounds backwards).

It's [annoying](https://nedbatchelder.com/blog/201003/whats_the_point_of_ospathcommonprefix.html) that Python does not have a better subdirectory detector -- maybe add a better function to `llnl.util.filesystem`.